feat: Allow total duration with max retries

tl-marco-tormento opened the pull request:
This PR fixes one main problem, which is the fact that using a total duration we are not able to set a limit to the number of retries that will be made inside that timeframe.

We added a new `build_with_total_retry_duration_and_max_retries` method to `ExponentialBackoffBuilder` that calculates the number of retries that should be made if jitter `1.0` was applied (which means no jitter).
Since jitter can be `0..1`, the calculated number of retries is effectively the max number of retries that can be made during that duration.

In `should_retry` we now check if either total duration or max attempts are exceeded.

We also made the exponential's base configurable: this will allow users to exponentially delay the retries more from one another.
The default remains `2`, so this isn't a breaking change.

ThomWright commented on src/policies/exponential_backoff.rs:
````diff
@@ -244,8 +244,74 @@ impl ExponentialBackoffBuilder {
+    /// exponential_backoff_timed.for_task_started_at(started_at)
+    ///     .should_retry(18); // RetryDecision::DoNotRetry
+    /// ```
+    pub fn build_with_total_retry_duration_and_max_retries(
````
I'm assuming this implements this:

![image](https://github.com/TrueLayer/retry-policies/assets/1884486/29c11b03-7b06-44b1-b243-314f3704c61f)

Rather than this:

![image](https://github.com/TrueLayer/retry-policies/assets/1884486/ad31c245-a145-44da-8a5d-ae8d9f4c83de)

If so, could we make this clear?

ThomWright replied:
Also, what are your thoughts on the effect of how long tasks take?

![image](https://github.com/TrueLayer/retry-policies/assets/1884486/9410cb4a-5722-4ee4-82e1-722d104ec9e7)

tl-marco-tormento replied:
> I'm assuming this implements this:

Exactly, I will make it clear.

ThomWright replied:
Cool, yeah that makes sense. I wasn't really sure what I was asking either! I figured it would work fine.

I was just aware that my mental model was originally assuming that task duration was negligible compared with retry intervals, which is often not true.

ThomWright reviewed (commented):
I'm wondering if there's a more general-purpose API design here.

I'm thinking about:

1. What if we want to set these independently?
2. What if we want to calculate duration from max retries? Is that a sensible option?

For 1, what do you think about an API like:

```rust
let exponential_backoff_timed = ExponentialBackoff::builder()
    .retry_bounds(Duration::from_secs(1), Duration::from_secs(6 * 60 * 60))
    .max_retries(10)
    .build_with_total_retry_duration(Duration::from_secs(24 * 60 * 60));
```

or:

```rust
ExponentialBackoff::builder()
    .build_with_max_retries(5)
    .with_total_retry_duration(Duration::from_secs(24 * 60 * 60));
```

Not sure about 2 yet, but something like this:

![image](https://github.com/TrueLayer/retry-policies/assets/1884486/b7141757-6382-4d20-8e9c-a1c9d26acff0)

I might be overcomplicating it, but keen to try to find a good general-purpose API which can cover these cases, instead of committing to specific-purpose APIs.

tl-marco-tormento commented:
> I'm thinking about:
> 
> 1. What if we want to set these independently?

What if we provide this API:
```rust
fn build_with_total_retry_duration(
        self,
        total_duration: Duration,
        max_retries: MaxRetries
    ) -> ExponentialBackoffTimed;

enum MaxRetries {
       Unbounded,
       Default,
       Fixed(u16)
}
```

User could:
- select `total_duration` and `MaxRetries::Unbounded` => do any number of retries inside `total_duration`
- select `total_duration` and `MaxRetries::Default` => we calculate max number of retries using `total_duration` and enforce whatever comes first
- select `total_duration` and `MaxRetries::Fixed(max_retries)` => we enforce whatever comes first, `max_retries` is capped at the number we calculate using `total_duration` anyway

> 2. What if we want to calculate duration from max retries? Is that a sensible option?
I think that should be the default when using the `build_with_max_retries` API, because I can't think of a case where I would want to set `max_retries` without capping the `total_duration`... do you have any example?

tl-marco-tormento commented on src/policies/exponential_backoff.rs:
```diff
@@ -24,9 +24,9 @@ pub struct ExponentialBackoff {
     /// Maximum duration the retries can continue for, after which retries will stop.
-    max_total_retry_duration: Duration,
+    pub max_total_retry_duration: Duration,
 
-    backoff: ExponentialBackoff,
+    pub backoff: ExponentialBackoff,
```
I did this to be able to assert max number of retries in the rust doc [here](https://github.com/TrueLayer/retry-policies/pull/15/files#diff-375bf27a13529254fcd25d253dbe06cb1115c1238ccea7f1778c40befdf82fb9R266) tbh.

eopb replied:
If it only needs to be read-only, could this be better off as a method?

eopb commented on src/policies/exponential_backoff.rs:
```diff
@@ -244,8 +249,85 @@ impl ExponentialBackoffBuilder {
+    ///     .checked_sub_signed(chrono::Duration::seconds(25 * 60 * 60))
+    ///     .unwrap();
+    ///
+    /// exponential_backoff_timed.for_task_started_at(started_at)
+    ///     .should_retry(0); // RetryDecision::DoNotRetry
```
Can we `assert!` here rather than just comment what the response should be?

ThomWright commented on src/policies/exponential_backoff.rs:
```diff
@@ -244,8 +249,86 @@ impl ExponentialBackoffBuilder {
+        let mut max_n_retries = None;
+
+        let delays = (0u32..).map(|n| {
+            min(
+                self.max_retry_interval,
+                self.min_retry_interval * 2_u32.checked_pow(n).unwrap_or(u32::MAX),
+            )
```
Nit: we could de-dupe this logic so we don't miss it when allowing us to change the base (currently fixed at 2_u32).
